src/algorithms/piecewise_linear_stretching.jl

```diff
@@ -0,0 +1,179 @@
+One subsequently constructs a concomitant new set of disjoint intervals
+```math
+I_{1}^{′} = [a_1,a_2], I_{2}^{′} = (a_2, a_3], \\ldots, I_{m}^{′} = (a_m,a_{m+1}]
+```
+such that ``\\sum_j |I_{j}^{′} | = | I^{′} |``, where ``I^{′} = [a,b]``. Finally, one introduces a
+sequence of mappings ``f_j : I_j \\to I_{j}^{′}`` ``(j = 1, \\ldots, m)`` given by
+```math
+f_j(x) = (x-A_{j}) \\frac{a_{j+1}-a_j}{A_{j+1}-A_{j}} + a_j.
+```
+which linearly map intensities from the interval ``I_j`` to the interval ``I_{j}^{′}`` .
+
+# Options
+
+Various options for the parameters of the `adjust_histogram` and
+`PiecewiseLinearStretching` type are described in more detail below.
+
+## Choices for `img`
+
+The function can handle a variety of input types. The returned image depends on the input
+type.
+
+For colored images, the input is converted to the [YIQ](https://en.wikipedia.org/wiki/YIQ)
+type and the intensities of the Y channel are stretched to the specified range. The modified
+Y channel is then combined with the I and Q channels and the resulting image converted to
+the same type as the input.
+
+## Choices for `src_intervals` 
+
+The `src_intervals` must be specified as an `ntuple` of pairs. For example,
+`src_intervals = (0.0 => 0.2, 0.2 => 0.8, 0.8 => 1.0)` partitions the unit
+interval into three subintervals, [0.0, 0.2], (0.2, 0.8] and (0.8, 1.0]. 
+
+To specify a single interval you must include a trailing comma, e.g.
+`src_intervals = (0.0 => 1.0,)`. 
+
+The start and end of a subinterval in `src_intervals` cannot be equal. For example,
+`src_intervals = (0.0 => 0.5, 0.5 => 0.5, 0.5 => 1.0)` will result in an `ArgumentError`
+because of the pair `0.5 => 0.5`.
+
+## Choices for `dst_intervals` 
+
+The `dst_intervals` must be specified as an `ntuple` of pairs and there needs to be a
+corresponding pair for each pair in `src_intervals`.
+    
+Unlike `src_intervals`, the start and end of an interval in `src_intervals` can be equal.
+For example, the combination `src_intervals = (0.0 => 0.5, 0.5 => 1.0)` and `dst_intervals =
+(0.0 => 0.0 , 1.0 => 1.0)` will produce a binary image. 
+
+Care must be taken to ensure that the subintervals specified in `dst_intervals` are
+representable within the type of image on which the piecewise linear stretching will be
+applied. For example, if you specify a negative output interval like `dst_intervals = (1.0
+=> -1.0),` and try to apply the transformation on an image with type `Gray{N0f8}`, you will
+receive a DomainError. 
+
+## Saturating pixels
+
+If the specified `src_intervals` don't span the entire range of intensities in an image,
+then intensities that fall outside `src_intervals` are automatically set to the boundary
+values of the `dst_intervals`. For example, given the combination `src_intervals = (0.1 =>
+0.9, )` and `dst_intervals = (0.0 => 1.0)`, intensities below `0.1` are set to zero, and
+intensities above `0.9` are set to one. 

```

Comment:
Perhaps IntervalSets.jl is the best option for the implementation, with a view to providing a "knots-based" constructor for convenience as you have both pointed out. I'm not sure that a solely knots-based representation is sufficiently general. For example, one may want to slightly brighten the dark parts of the image, and slightly darken the bright part of the image, leaving the rest unchanged. 

I can imagine specifying this as:
```julia
PiecewiseLinearStretching(0.0..0.1 => 0.0..0.2, 0.1..0.8 => 0.1..0.8, 0.8..1.0 => 0.7..1.0)
```
Unless I misunderstood, the knots-based variant wouldn't work because the source knots are: `0, 0.1, 0.8, 1.0` but the destination knots are `0, 0.2, 0.1, 0.8, 0.7, 1.0` (different lengths).

A potential different way of specifying this might be to leave out the "identity" mapping `0.1..0.8 => 0.1..0.8`
```julia
PiecewiseLinearStretching(0.0..0.1 => 0.0..0.2, 0.8..1.0 => 0.7..1.0)
```

However, as Johnny has quite pointed out, I currently don't handle the case where there are "gaps in the middle". Part of the issue is that I have introduced ambiguity in what it means to not specify an interval. In one case, pixels get "saturated" to the boundary values, whereas in other cases it is an implicit "identity" mapping. 

Maybe the best way to handle this is to introduce a keyword argument that allows one to specify whether the edges should be saturated, and then internally explicitly specify the interval mapping that results in the saturation.